Support multiple OpenBSD versions

Blaisorblade opened the pull request:
Sketch a different setup for OpenBSD: use OpenBSD release as ghc-build identifier. See comments for details.

Should ideally be merged before https://github.com/commercialhaskell/stack/pull/3184.

Blaisorblade commented:
Testers can try out these changes locally via https://docs.haskellstack.org/en/stable/yaml_configuration/#setup-info.

Blaisorblade commented on stack/stack-setup-2.yaml:
```diff
@@ -512,6 +512,30 @@ ghc:
+        8.0.2:
+            url: "https://github.com/commercialhaskell/ghc/releases/download/ghc-8.0.2-release/ghc-8.0.2-x86_64-unknown-openbsd.tar.xz"
+            content-length: 100054832
+            sha1: 426b01e6ac72b5e039eb9626649506bc47eb9510
```
This block could in theory be merged already (though it'd be unused and might bitrot) and should work once https://github.com/commercialhaskell/stack/pull/3184 is in.

Blaisorblade commented on stack/stack-setup-2.yaml:
```diff
@@ -512,6 +512,30 @@ ghc:
+            # provided by https://github.com/mrijkeboer
+            url: "https://s3.amazonaws.com/download.fpcomplete.com/ghc/ghc-7.10.3-openbsd-x86_64-current.tar.bz2"
+            content-length: 135191335
+            sha1: 35f8b595434fcbfa2cb6fc0e9561a5a4d87382a5
```
This block should probably be dropped, but it seems so should be (maybe?) all bindists compiled on current. Not sure though.

Blaisorblade commented:
Ping @borsboom. I propose to merge this (which should have no effect on existing users), and then merge https://github.com/commercialhaskell/stack/pull/3184.

I should try adding GHC 8.2.1 bindists for OpenBSD 6.0 and 6.1 to fix #27, using this same scheme, but I'm on a hurry right now: https://www.haskell.org/ghc/download_ghc_8_2_1.html#openbsd_61_x86_64

borsboom commented:
This sounds pretty reasonable.  Do we need the `maj6-min1` formatting, or could we just use something simpler like `6_1`?

We may need exactly the same approach for FreeBSD (where the GHC build team is now releasing different binaries for 10.3 and 11).

borsboom commented:
Looks like this PR is missing GHC 8.2.1.  Can you add that?

Blaisorblade commented:
> Looks like this PR is missing GHC 8.2.1. Can you add that?

Right—rebased and fixed. I've also added 8.2.1 for OpenBSD 6.1 (using the new scheme) to fix #27. I've had to use the original URL on `downloads.haskell.org`—can you mirror it as needed?

borsboom commented:
I've mirrored the bindist to https://github.com/commercialhaskell/ghc/releases/download/ghc-8.2.1-release/ghc-8.2.1-x86_64-openbsd61-openbsd.tar.xz.  Can you update?

Blaisorblade commented:
Done! Can we merge?

borsboom commented:
Done.  I'll give you commit bits on this repo and the GHC mirror repo so you can make further adjustments yourself!
